voice/src/main/kotlin/gateway/VoiceGateway.kt

```diff
@@ -83,6 +83,34 @@ interface VoiceGateway : CoroutineScope {
     }
 }
 
+
+/**
+ * Logger used to report throwables caught in [VoiceGateway.on].
+ */
+@PublishedApi
+internal val voiceGatewayOnLogger = KotlinLogging.logger("Gateway.on")
+
+/**
+ * Convenience method that will invoke the [consumer] on every event [T] created by [VoiceGateway.events].
+ *
+ * The events are buffered in an [unlimited][Channel.UNLIMITED] [buffer][Flow.buffer] and
+ * [launched][CoroutineScope.launch] in the supplied [scope], which is [VoiceGateway] by default.
+ * Each event will be [launched][CoroutineScope.launch] inside the [scope] separately and
+ * any thrown [Throwable] will be caught and logged.
+ *
+ * The returned [Job] is a reference to the created coroutine, call [Job.cancel] to cancel the processing of any further
+ * events for this [consumer].
+ */
+@KordVoice
+inline fun <reified T : VoiceEvent> VoiceGateway.on(
+    scope: CoroutineScope = this,
+    crossinline consumer: suspend T.() -> Unit
+): Job {
+    return this.events.buffer(Channel.UNLIMITED).filterIsInstance<T>().onEach {
+        launch { it.runCatching { it.consumer() }.onFailure(voiceGatewayOnLogger::error) }
+    }.launchIn(scope)
+}
```

Comment:
This should not launch inside `onEach` to make sure that changes to the `MutableMap` in `Streams` are sequential (see @BartArys comment). Other solution would be to not use `VoiceGateway.on()` in `Streams`.

Reply:
This function should stay the same, as it would be confusing that the VoiceGateway.on functions differently from Gateway.on
Though I now understand what you meant by not using launch, I did not realize it was used here :p

Reply:
```kotlin
connection.voiceGateway.events.buffer(Channel.UNLIMITED).filterIsInstance<Speaking>().onEach { speaking ->
    // ...
}.launchIn(this)
```
in `init` for `Streams` should do the trick then